### PR TITLE
Notify about Track completion and failure

### PR DESCRIPTION
**Pull request by UselessToucan**
`Track` currently doesn't signal its completion. This issue is fixed here.

### CODE REVIEW

**Comment by UselessToucan** [comment]
Yes, I tested it in game. Stopping preview tracks and switching to the next track in the music controller works ok. Here is my [game PR](https://github.com/ppy/osu/pull/3732).

**Comment by UselessToucan** [comment]
But I did not test it with looping tracks. I will do it today.

**Comment by peppy** [comment]
If these are to be at a `Track` level then they should be implemented by all track types (ie `TrackVirtual`) else you are going to regress anything which uses them.

File: osu.Framework/Audio/Track/Track.cs [hunk 1]
```diff
@@ -10,6 +10,9 @@ namespace osu.Framework.Audio.Track
 {
     public abstract class Track : AdjustableAudioComponent, IAdjustableClock
     {
+        public virtual event Action Completed;
```
Comment (peppy) [inline osu.Framework/Audio/Track/Track.cs]: why virtual?

File: osu.Framework/Audio/Track/TrackBass.cs [hunk 2]
```diff
@@ -17,6 +17,9 @@ public class TrackBass : Track, IBassAudio, IHasPitchAdjust
     {
         private AsyncBufferStream dataStream;
 
+        public override event Action Completed;
```
Comment (peppy) [inline osu.Framework/Audio/Track/TrackBass.cs]: why overridden?

File: osu.Framework/Audio/Track/TrackBass.cs [hunk 3]
```diff
@@ -95,6 +95,14 @@ public TrackBass(Stream data, bool quick = false)
                     initialFrequency = frequency;
                     bitrate = (int)Bass.ChannelGetAttribute(activeStream, ChannelAttribute.Bitrate);
 
+                    // Init Events
+                    Bass.ChannelSetSync(activeStream, SyncFlags.Stop, 0, (a, b, c, d) => RaiseFailed());
```
Comment (peppy) [inline osu.Framework/Audio/Track/TrackBass.cs]: this may require a local reference to work on iOS

Comment (peppy) [reply osu.Framework/Audio/Track/TrackBass.cs]: @swoolcock can you test this if you have a chance?

Comment (swoolcock) [reply osu.Framework/Audio/Track/TrackBass.cs]: It will absolutely require an iOS fix, since the Bass procedures must be static references.

Comment (swoolcock) [reply osu.Framework/Audio/Track/TrackBass.cs]: It's part of the reason this fails: https://github.com/ManagedBass/ManagedBass/blob/master/src/Bass/Shared/GCPin.cs#L35

Comment (UselessToucan) [reply osu.Framework/Audio/Track/TrackBass.cs]: Do I understand correctly that `(a, b, c, d) => RaiseFailed()` should be similar to [this](https://github.com/ppy/osu-framework/blob/master/osu.Framework.iOS/Audio/IOSSampleBass.cs#L49)? I mean, it should be implemented as a `static` method with `[MonoPInvokeCallback(typeof(SyncProcedure))]` attribute.

Comment (swoolcock) [reply osu.Framework/Audio/Track/TrackBass.cs]: I don't see how their `GCPin.CreateStreamHelper` has ever worked on iOS, yet they have an iOS csproj.

**Comment by UselessToucan** [comment]
This PR will be reworked because ['BassCallback pattern' was introduced](https://github.com/ppy/osu-framework/commit/dbb80a5d2e88c16f6348585e074688775d09297f#diff-5f2c28017af611de3534a93a80cabded).

**Comment by swoolcock** [comment]
@UselessToucan the factory pattern in that commit was reverted in favour of an easier callback using a local definition of `MonoPInvokeCallbackAttribute`.

You should be able to do the sync implementation as such:

```csharp
// store and dispose these
SyncCallback stopCallback = new SyncCallback((a, b, c, d) => RaiseFailed());
SyncCallback endCallback = new SyncCallback((a, b, c, d) =>
{
    if (!Looping)
        RaiseCompleted();
});

Bass.ChannelSetSync(activeStream, SyncFlags.Stop, 0, stopCallback.Callback, stopCallback.Handle);
Bass.ChannelSetSync(activeStream, SyncFlags.End, 0, endCallback.Callback, endCallback.Handle);
```

**Pull request merged**